braindecode/datasets/base.py

```diff
@@ -46,53 +64,81 @@ def __init__(self, raw, description=None, target_name=None):
 
     def __getitem__(self, index):
-        return self.raw[:, index][0], self.target
+        X, y = self.raw[:, index][0], self.target
+        if self.transform is not None:
+            X = self.transform(X)
+        return X, y
 
     def __len__(self):
         return len(self.raw)
 
+    @property
+    def transform(self):
+        return self._transform
 
-def _create_description(description):
-    if description is not None:
-        if (not isinstance(description, pd.Series)
-                and not isinstance(description, dict)):
-            raise ValueError(f"'{description}' has to be either a "
-                             f"pandas.Series or a dict")
-        if isinstance(description, dict):
-            description = pd.Series(description)
-    return description
+    @transform.setter
+    def transform(self, value):
+        if value is not None and not callable(value):
+            raise ValueError('Transform needs to be a callable.')
+        self._transform = value
 
 
 class WindowsDataset(BaseDataset):
-    """Applies a windower to a base dataset.
+    """Returns windows from an mne.Epochs object along with a target.
+
+    Dataset which serves windows from an mne.Epochs object along with their
+    target and additional information. The `metadata` attribute of the Epochs
+    object must contain a column called `target`, which will be used to return
+    the target that corresponds to a window. Additional columns
+    `i_window_in_trial`, `i_start_in_trial`, `i_stop_in_trial` are also
+    required to serve information about the windowing (e.g., useful for cropped
+    training).
+    See `braindecode.datautil.windowers` to directly create a `WindowsDataset`
+    from a `BaseDataset` object.
 
     Parameters
     ----------
-    windows: mne.Epochs
-        windows obtained through the application of a windower to a
-        BaseDataset
-    description: dict | pandas.Series | None
-        holds additional info about the windows
+    windows : mne.Epochs
+        Windows obtained through the application of a windower to a BaseDataset
+        (see `braindecode.datautil.windowers`).
+    description : dict | pandas.Series | None
+        Holds additional info about the windows.
+    transform : callable | None
+        On-the-fly transform applied to a window before it is returned.
     """
-    def __init__(self, windows, description=None):
+    def __init__(self, windows, description=None, transform=None):
         self.windows = windows
         self.description = _create_description(description)
-        self.y = np.array(self.windows.metadata.loc[:, 'target'])
-        self.crop_inds = np.array(self.windows.metadata.loc[:,
-                                  ['i_window_in_trial', 'i_start_in_trial',
-                                   'i_stop_in_trial']])
+        self.transform = transform
+
+        self.y = self.windows.metadata.loc[:, 'target'].values
```

Comment:
This is purely a personal preference, I think using the object's dedicated method is cleaner than calling an external function on the object. Not sure about here but sometimes this could also end up being more efficient if the method was designed in a certain way. I don't mind if you prefer the old way though!

Reply:
In general, seems fine to me. However just checked https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.values.html and it says:
![image](https://user-images.githubusercontent.com/1178948/114785209-9e2a2600-9d7c-11eb-8f8b-5399df45c68e.png)

So since we are switching anyways, let's switch to that how do you think?

Reply:
HM or is this series? https://pandas.pydata.org/docs/reference/api/pandas.Series.values.html 
Then this may apply:
![image](https://user-images.githubusercontent.com/1178948/114785340-dcbfe080-9d7c-11eb-96d1-c93957c5f4eb.png)

Reply:
```suggestion
        self.y = self.windows.metadata.loc[:, 'target'].to_numpy()
```

Reply:
Nice, didn't know about that! Works for me.